File: coding exercise/template.py
#!/usr/bin/env python3
import collections
import heapq
import math
import bisect
from typing import List, Optional, Dict, Set, Tuple

# ================= LeetCode Common Data Structures =================

class ListNode:
    """Definition for singly-linked list."""
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __repr__(self):
        curr = self
        vals = []
        while curr:
            vals.append(str(curr.val))
            curr = curr.next
        return " -> ".join(vals) if vals else "Empty List"


class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __repr__(self):
        return f"TreeNode({self.val})"


# ================= Helper Functions for Local Testing =================

def list_to_linked_list(nums: List[int]) -> Optional[ListNode]:
    """Converts a standard Python list to a ListNode linked list."""
    dummy = ListNode(0)
    curr = dummy
    for num in nums:
        curr.next = ListNode(num)
        curr = curr.next
    return dummy.next


def linked_list_to_list(head: Optional[ListNode]) -> List[int]:
    """Converts a ListNode linked list back to a standard Python list."""
    res = []
    curr = head
    while curr:
        res.append(curr.val)
        curr = curr.next
    return res


def list_to_binary_tree(arr: List[Optional[int]]) -> Optional[TreeNode]:
    """
    Converts a LeetCode-style level-order list (including None for null nodes)
    into a TreeNode binary tree.
    Example: list_to_binary_tree([1, None, 2, 3])
    """
    if not arr or arr[0] is None:
        return None
    
    root = TreeNode(arr[0])
    queue = collections.deque([root])
    i = 1
    
    while queue and i < len(arr):
        curr = queue.popleft()
        
        # Left child
        if i < len(arr):
            if arr[i] is not None:
                curr.left = TreeNode(arr[i])
                queue.append(curr.left)
            i += 1
            
        # Right child
        if i < len(arr):
            if arr[i] is not None:
                curr.right = TreeNode(arr[i])
                queue.append(curr.right)
            i += 1
            
    return root


def binary_tree_to_list(root: Optional[TreeNode]) -> List[Optional[int]]:
    """Converts a TreeNode binary tree back to a LeetCode level-order list."""
    if not root:
        return []
        
    res = []
    queue = collections.deque([root])
    
    while queue:
        curr = queue.popleft()
        if curr:
            res.append(curr.val)
            queue.append(curr.left)
            queue.append(curr.right)
        else:
            res.append(None)
            
    # Trim trailing Nones to match LeetCode output style
    while res and res[-1] is None:
        res.pop()
        
    return res


def print_tree(root: Optional[TreeNode], indent: str = "", is_left: bool = True) -> None:
    """Helper to visualize tree structure in console."""
    if root is None:
        return
    
    if root.right:
        print_tree(root.right, indent + ("│   " if is_left else "    "), False)
        
    print(indent + ("└── " if is_left else "┌── ") + str(root.val))
    
    if root.left:
        print_tree(root.left, indent + ("    " if is_left else "│   "), True)


# ================= Solution Class =================

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Example solution (Two Sum)
        seen = {}
        for i, num in enumerate(nums):
            diff = target - num
            if diff in seen:
                return [seen[diff], i]
            seen[num] = i
        return []


# ================= Local Runner =================

if __name__ == '__main__':
    import sys
    
    sol = Solution()
    
    # Automatically find the public solver method on Solution (ignoring helper/private methods)
    methods = [m for m in dir(sol) if not m.startswith('_') and callable(getattr(sol, m))]
    if not methods:
        print("Error: No public solver method found in Solution class!")
        sys.exit(1)
        
    method_name = methods[0]
    solve = getattr(sol, method_name)
    print(f"Running tests for: Solution.{method_name}()\n")

    # Define your test cases: [ (inputs, expected_output), ... ]
    # - If your method takes a single argument, pass it directly or wrap in a tuple: (arg,)
    # - If your method takes multiple arguments, pass them as a tuple: (arg1, arg2, ...)
    test_cases = [
        # Test Case 1: nums = [2, 7, 11, 15], target = 9 -> expected = [0, 1]
        ( ([2, 7, 11, 15], 9), [0, 1] ),
    ]

    for i, (args, expected) in enumerate(test_cases, 1):
        print(f"--- Test Case {i} ---")
        # Ensure args is a tuple for dynamic unpacking
        if not isinstance(args, tuple):
            args = (args,)
            
        # Try to print input arguments nicely
        readable_input = args[0] if len(args) == 1 else args
        print(f"Input:    {readable_input}")
        
        result = solve(*args)
        
        print(f"Result:   {result}")
        print(f"Expected: {expected}")
        
        assert result == expected, f"Test Case {i} Failed!"
        print(f"Test Case {i} Passed!\n")
        
    print("All tests run successfully!")

File: coding exercise/template.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <queue>
#include <stack>
#include <algorithm>
#include <numeric>
#include <climits>
#include <cmath>
#include <sstream>
#include <cassert>

// ================= LeetCode Common Data Structures =================

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// ================= Helper Functions for Local Testing =================

// Linked List helpers
ListNode* listToLinkedList(const std::vector<int>& nums) {
    ListNode dummy(0);
    ListNode* curr = &dummy;
    for (int num : nums) {
        curr->next = new ListNode(num);
        curr = curr->next;
    }
    return dummy.next;
}

std::vector<int> linkedListToList(ListNode* head) {
    std::vector<int> res;
    ListNode* curr = head;
    while (curr) {
        res.push_back(curr->val);
        curr = curr->next;
    }
    return res;
}

void freeLinkedList(ListNode* head) {
    while (head) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

void printList(ListNode* head) {
    ListNode* curr = head;
    while (curr) {
        std::cout << curr->val << (curr->next ? " -> " : "");
        curr = curr->next;
    }
    std::cout << std::endl;
}

// Binary Tree helpers
TreeNode* listToBinaryTree(const std::vector<std::string>& arr) {
    if (arr.empty() || arr[0] == "null" || arr[0] == "None" || arr[0] == "") {
        return nullptr;
    }
    
    TreeNode* root = new TreeNode(std::stoi(arr[0]));
    std::queue<TreeNode*> q;
    q.push(root);
    size_t i = 1;
    
    while (!q.empty() && i < arr.size()) {
        TreeNode* curr = q.front();
        q.pop();
        
        // Left child
        if (i < arr.size()) {
            if (arr[i] != "null" && arr[i] != "None" && arr[i] != "") {
                curr->left = new TreeNode(std::stoi(arr[i]));
                q.push(curr->left);
            }
            i++;
        }
        
        // Right child
        if (i < arr.size()) {
            if (arr[i] != "null" && arr[i] != "None" && arr[i] != "") {
                curr->right = new TreeNode(std::stoi(arr[i]));
                q.push(curr->right);
            }
            i++;
        }
    }
    return root;
}

std::vector<std::string> binaryTreeToList(TreeNode* root) {
    std::vector<std::string> res;
    if (!root) return res;
    
    std::queue<TreeNode*> q;
    q.push(root);
    
    while (!q.empty()) {
        TreeNode* curr = q.front();
        q.pop();
        
        if (curr) {
            res.push_back(std::to_string(curr->val));
            q.push(curr->left);
            q.push(curr->right);
        } else {
            res.push_back("null");
        }
    }
    
    // Trim trailing "null"s
    while (!res.empty() && res.back() == "null") {
        res.pop_back();
    }
    return res;
}

void freeBinaryTree(TreeNode* root) {
    if (!root) return;
    freeBinaryTree(root->left);
    freeBinaryTree(root->right);
    delete root;
}

void printTree(TreeNode* root, std::string indent = "", bool isLeft = true) {
    if (!root) return;
    
    if (root->right) {
        printTree(root->right, indent + (isLeft ? "│   " : "    "), false);
    }
    
    std::cout << indent << (isLeft ? "└── " : "┌── ") << root->val << std::endl;
    
    if (root->left) {
        printTree(root->left, indent + (isLeft ? "    " : "│   "), true);
    }
}

// Helper to print standard vectors
template <typename T>
void printVector(const std::vector<T>& vec) {
    std::cout << "[";
    for (size_t i = 0; i < vec.size(); ++i) {
        std::cout << vec[i] << (i + 1 < vec.size() ? ", " : "");
    }
    std::cout << "]" << std::endl;
}


// ================= Solution Class =================

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        // Example solution (Two Sum)
        std::unordered_map<int, int> seen;
        for (int i = 0; i < nums.size(); ++i) {
            int diff = target - nums[i];
            if (seen.count(diff)) {
                return {seen[diff], i};
            }
            seen[nums[i]] = i;
        }
        return {};
    }
};


// ================= Local Runner =================

int main() {
    Solution sol;
    
    // Test Case 1
    {
        std::cout << "--- Test Case 1 ---" << std::endl;
        std::vector<int> nums = {2, 7, 11, 15};
        int target = 9;
        std::vector<int> expected = {0, 1};
        
        std::vector<int> result = sol.twoSum(nums, target);
        
        assert(result == expected && "Test Case 1 Failed!");
        std::cout << "Test Case 1 Passed!\n" << std::endl;
    }
    
    // Test Case 2 (Template placeholder)
    /*
    {
        std::cout << "--- Test Case 2 ---" << std::endl;
        std::vector<int> nums = {3, 2, 4};
        int target = 6;
        std::vector<int> expected = {1, 2};
        
        std::vector<int> result = sol.twoSum(nums, target);
        
        assert(result == expected && "Test Case 2 Failed!");
        std::cout << "Test Case 2 Passed!\n" << std::endl;
    }
    */
    
    std::cout << "All tests run successfully!" << std::endl;
    return 0;
}
